src/libstd/path.rs

```diff
@@ -4020,6 +4106,165 @@ mod tests {
         assert_eq!(format!("a{:#<5}b", Path::new("a").display()), "aa####b");
     }
 
+    #[test]
+    pub fn test_clean() {
+        macro_rules! tc(
+            ($path:expr, $expected:expr) => ( {
+                let mut actual = PathBuf::from($path).clean();
+                assert!(actual.to_str() == Some($expected),
+                        "cleaning {:?}: Expected {:?}, got {:?}",
+                        $path, $expected,
+                        actual.to_str().unwrap());
+            });
+        );
+
+        if cfg!(unix) {
+            tc!("", ".");
+            tc!("/", "/");
+            tc!("//", "/");  /* Double-slash root is a separate entity in POSIX,
+                            but in Rust we treat it as a normal root slash. */
+            tc!("foo", "foo");
+            tc!(".", ".");
+            tc!("..", "..");
+            tc!(".foo", ".foo");
+            tc!("..foo", "..foo");
+            tc!("/foo", "/foo");
+            tc!("//foo", "/foo");
+            tc!("./foo/", "foo");
+            tc!("../foo/", "../foo");
+            tc!("/foo/bar", "/foo/bar");
+            tc!("foo/bar", "foo/bar");
+            tc!("foo/.", "foo");
+            tc!("foo//bar", "foo/bar");
+            tc!("./foo//bar//", "foo/bar");
+
+            tc!("foo/bar/baz/..", "foo/bar");
+            tc!("foo/bar/baz/../", "foo/bar");
+            tc!("foo/bar/baz/../..", "foo");
+            tc!("foo/bar/baz/../../..", ".");
+            tc!("foo/bar/baz/../../../..", "..");
+            tc!("foo/bar/baz/../../../../..", "../..");
+            tc!("/foo/bar/baz/../../../../..", "/");
+            tc!("foo/../bar/../baz/../", ".");
+            tc!("/.", "/");
+            tc!("/..", "/");
+            tc!("/../../", "/");
+        } else {
+            // Drive-absolute paths.
+            tc!(r#"X:\ABC\DEF"#, r#"X:\ABC\DEF"#);
```

Comment:
Nit pick. Unless your path contains a `"` you could just lose the `#`s

```rust
tc!(r"X:\ABC\DEF", r"X:\ABC\DEF");
```

I think this is much less noisy.